Pathname: add write_append support

DomT4 opened the pull request:
I don't know how popular an idea this'll be, but since it didn't take long to throw together _(I went to get food, this didn't take three hours :stuck_out_tongue_winking_eye:)_, I thought a PR would be nicer than me occupying the team chat like a persistent protest movement again.

* Blocks writing of new files via accidental typos, etc, which the normal `open("blah", "a")` doesn't.
* Where files don't exist they should _ideally_ be using the `(buildpath/"dog").write` style instead of `open("blah", "a")` already.
* It's a bit less cluttered looking if you need several writes to different files in the formula, IMO.

DomT4 commented:
Jenkins failure looks like PHP contamination. Back in the queue it goes.

xu-cheng commented on Library/Homebrew/extend/pathname.rb:
```diff
@@ -131,6 +131,12 @@ def write(content, *open_args)
+  # Only appends to a file that is already created.
+  def write_append(content, *open_args)
+    raise "Cannot append file that doesn't exist: #{self}" unless exist?
+    open("a", *open_args) { |f| f.puts(content) }
```
I think it should be `f.write` instead of `f.puts` here, as it's surprise to silently append newline. The caller should explicitly pass the newline character if needed.

ilovezfs replied:
What is the justification for requiring file existence? Not saying it's a bad idea, but wondering what the rationale is for not making it equivalent to a shell `>>` append redirection?

UniqMartin replied:
> What is the justification for requiring file existence?

To quote from the top comment:

> - Blocks writing of new files via accidental typos, etc, which the normal `open("blah", "a")` doesn't.

DomT4 replied:
Aye. We also already have `(buildpath/"dogs").write "Are the best"` if you need to write to an empty file.

DomT4 replied:
> I think it should be f.write instead of f.puts here, as it's surprise to silently append newline. The caller should explicitly pass the newline character if needed.

I think `f.puts` is consistent with current usage and expectation, but I guess I don't have a particularly strong objection if nobody else does.

UniqMartin replied:
While I can understand the reason for using `f.puts`, I would also prefer to see it replaced with `f.write`. I think making the interface in its interpretation of the `content` argument consistent with `Pathname#write` (implemented directly above) trumps the minor inconvenience of having to add a few `\n` in some places.

DomT4 replied:
Fair enough. I'll rewrite to adjust for that sometime tonight/tomorrow.

tdsmith replied:
Since POSIX-y text files should always end with a newline I think requiring the user to add `\n` just makes the API harder to use correctly. Would we feel differently if it were called something like `append_lines` instead?

mikemcquaid replied:
@tdsmith makes a good point. `append_lines` makes sense to me, here.

UniqMartin replied:
Taking @tdsmith's comment into account my view has shifted: Because it was named `write_append` I expected it to behave more like `write`, but naming it `append_lines` makes that less of an issue. The core of the issue is maybe that sometimes it's called with a single string, where it's desirable to have that being auto-terminated by a newline, and sometimes it's called with a heredoc that already contains a newline at the end. But from the method's point of view it's all just strings. (Sorry if these random thoughts that are somewhat useless. I wanted to share them anyway.)

DomT4 replied:
I'm not enamoured with the alternative name, but I'll take it if it's the only thing holding this up at this point.

UniqMartin commented on Library/Formula/plan9port.rb:
```diff
@@ -17,15 +17,10 @@ class Plan9port < Formula
-      end
-    end
+    # Make OS X system fonts available to Plan 9
+    (buildpath/"LOCAL.config").write "FONTSRV=fontsrv" if build.with? "x11"
```
I assume this should have been `write_append` instead?

DomT4 replied:
Nope, this file doesn't exist until we create it. Discovered that last night when changing it to `write_append` encountered my raise because the file didn't exist :smile:.

UniqMartin replied:
Thanks for the clarification! I hadn't tested this myself and found this to be rather odd in the context of this PR. (Good thing there is this check for existence/non-existence of the target file that helps to avoid subtle mistakes like this one could have been!)

DomT4 commented:
Alright, fixed, rebased and pushed. If it passes CI I'll merge it shortly after, presuming everyone is happy with the rename having been the remaining issue.
